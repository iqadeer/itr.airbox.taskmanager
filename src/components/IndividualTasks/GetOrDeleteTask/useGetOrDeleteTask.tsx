import { ChangeEventHandler, useState, MouseEventHandler } from 'react';
import { AbxTask } from '../../../shared/models/AbxTask';
import { APIService } from '../../../api/ApiService';

export type AbxTaskGetResponse = {
  data: AbxTask[];
};

const useGetOrDeleteTask = () => {
  const [abxTaskId, setAbxTaskId] = useState<string>('');
  const [abxTaskIdError, setAbxTaskIdError] = useState<string>('');

  const handleChange: ChangeEventHandler<HTMLInputElement> = ({ target: { name, value } }) => {
    setAbxTaskId(value);

    const errorMessage = !value ? `${name} is required` : '';
    setAbxTaskIdError(errorMessage);
  };

  const isGetTaskStateValid = () => {
    setAbxTaskIdError(abxTaskId ? '' : 'abxTaskId is required');
    return !!abxTaskId;
  };

  const handleGet: MouseEventHandler<HTMLButtonElement> = async (e) => {
    e.preventDefault();
    if (isGetTaskStateValid()) {
      try {
        const result = await APIService.get<AbxTaskGetResponse>(`tasks/${abxTaskId}`, 'token');
        setSnackBar({ message: JSON.stringify(result.data), severity: 'success' });
        setOpenSnackBar(true);
        setAbxTaskId('');
        console.log(result.data);
      } catch (e) {
        setSnackBar({ message: `Error getting task: ${e}`, severity: 'error' });
        setOpenSnackBar(true);
      }
    }
  };

  const [openCancelDialog, setOpenCancelDialog] = useState<boolean>(false);

  const openOrCloseDialog = (open: boolean) => {
    if (open) {
      if (isGetTaskStateValid()) {
        setOpenCancelDialog(true);
      }
    } else {
      setOpenCancelDialog(false);
    }
  };

  const deleteTask = async () => {
    try {
      await APIService.delete(`tasks/${abxTaskId}`, 'token');
      setSnackBar({ message: 'Task deleted', severity: 'success' });
      setOpenSnackBar(true);
      setAbxTaskId('');
    } catch (e) {
      setSnackBar({ message: `Error deleting task: ${e}`, severity: 'error' });
      setOpenSnackBar(true);
    }
    setOpenCancelDialog(false);
  };

  const [openSnackBar, setOpenSnackBar] = useState<boolean>(false);
  const [snackBar, setSnackBar] = useState<{ message: string; severity: 'error' | 'success' }>({
    message: '',
    severity: 'success',
  });
  const closeSnackBar = (): void => {
    setOpenSnackBar(false);
  };

  return {
    deleteTask,
    openOrCloseDialog,
    openCancelDialog,
    handleGet,
    abxTaskIdError,
    handleChange,
    abxTaskId,
    openSnackBar,
    closeSnackBar,
    message: snackBar.message,
    severity: snackBar.severity,
  };
};

export default useGetOrDeleteTask;
