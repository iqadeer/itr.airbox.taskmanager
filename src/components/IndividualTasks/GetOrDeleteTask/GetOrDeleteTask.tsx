import React from 'react';
import {
  Grid,
  Button,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogContentText,
  DialogActions,
  TextField,
  Snackbar,
} from '@material-ui/core';
import useStyles from './useStyles';
import UpdateIcon from '@material-ui/icons/Update';
import DeleteIcon from '@material-ui/icons/Delete';
import MuiAlert, { AlertProps } from '@material-ui/lab/Alert';
import useGetOrDeleteTask from './useGetOrDeleteTask';

const Alert = (props: AlertProps) => {
  return <MuiAlert elevation={16} variant='filled' {...props} />;
};
const GetOrDeleteTask = () => {
  const classes = useStyles({});
  const {
    abxTaskId,
    abxTaskIdError,
    openOrCloseDialog,
    openCancelDialog,
    handleChange,
    handleGet,
    deleteTask,
    openSnackBar,
    closeSnackBar,
    message,
    severity,
  } = useGetOrDeleteTask();
  return (
    <>
      <form className={classes.form} noValidate>
        <Grid container justify='center' alignItems='center' spacing={2}>
          <Grid item xs={12} md={6}>
            <TextField
              autoFocus
              error={!!abxTaskIdError}
              placeholder='Airbox taskId'
              autoComplete='Airbox taskId'
              name='abxTaskId'
              variant='outlined'
              required
              fullWidth
              id='abxTaskId'
              value={abxTaskId}
              onChange={handleChange}
              label='Airbox taskId'
              helperText={abxTaskIdError}
            />
          </Grid>
        </Grid>

        <Grid container justify='center' alignItems='center'>
          <Grid item xs={12} md={6}>
            <Button
              type='submit'
              variant='contained'
              color='primary'
              startIcon={<UpdateIcon />}
              onClick={handleGet}
              id='getTask'
              className={classes.submit}>
              Get task
            </Button>
            <Button
              type='button'
              variant='outlined'
              color='primary'
              startIcon={<DeleteIcon />}
              onClick={() => openOrCloseDialog(true)}
              id='deleteTask'
              className={classes.cancel}>
              Delete task
            </Button>
          </Grid>
        </Grid>
      </form>
      <Dialog open={openCancelDialog} onClose={() => openOrCloseDialog(false)} aria-labelledby='form-dialog-title'>
        <DialogTitle id='form-dialog-title'>Confirm delete</DialogTitle>
        <DialogContent>
          <DialogContentText>Are you sure you want to delete Airbox task with id {abxTaskId}? </DialogContentText>
        </DialogContent>
        <DialogActions>
          <Button onClick={deleteTask} color='primary'>
            Confirm
          </Button>
          <Button onClick={() => openOrCloseDialog(false)} color='primary'>
            Cancel
          </Button>
        </DialogActions>
      </Dialog>
      <Snackbar
        autoHideDuration={3000}
        anchorOrigin={{ vertical: 'bottom', horizontal: 'center' }}
        key={`top,center`}
        open={openSnackBar}
        onClose={closeSnackBar}>
        <Alert onClose={closeSnackBar} severity={severity}>
          {message}
        </Alert>
      </Snackbar>
    </>
  );
};

export default GetOrDeleteTask;
