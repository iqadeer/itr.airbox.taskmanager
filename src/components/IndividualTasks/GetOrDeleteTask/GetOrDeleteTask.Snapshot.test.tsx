import React from 'react';
import renderer from 'react-test-renderer';
import GetOrDeleteTask from './GetOrDeleteTask';

it('should have correct button label', () => {
  const task = renderer.create(<GetOrDeleteTask></GetOrDeleteTask>);
  expect(task).toMatchSnapshot();
});
