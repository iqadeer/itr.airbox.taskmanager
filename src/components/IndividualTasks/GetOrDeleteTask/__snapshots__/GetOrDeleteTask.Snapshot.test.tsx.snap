// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should have correct button label 1`] = `
<form
  className="makeStyles-form-5"
  noValidate={true}
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink={false}
          htmlFor="abxTaskId"
          id="abxTaskId-label"
        >
          Airbox taskId
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="Airbox taskId"
            autoFocus={true}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="abxTaskId"
            name="abxTaskId"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Airbox taskId"
            required={true}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-10 MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="PrivateNotchedOutline-legendLabelled-12"
            >
              <span>
                Airbox taskId
                 *
              </span>
            </legend>
          </fieldset>
        </div>
        
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-6"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-8 MuiButton-containedPrimary"
        disabled={false}
        id="getTask"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M21 10.12h-6.78l2.74-2.82c-2.73-2.7-7.15-2.8-9.88-.1-2.73 2.71-2.73 7.08 0 9.79s7.15 2.71 9.88 0C18.32 15.65 19 14.08 19 12.1h2c0 1.98-.88 4.55-2.64 6.29-3.51 3.48-9.21 3.48-12.72 0-3.5-3.47-3.53-9.11-.02-12.58s9.14-3.47 12.65 0L21 3v7.12zM12.5 8v4.25l3.5 2.08-.72 1.21L11 13V8h1.5z"
              />
            </svg>
          </span>
          Get task
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-cancel-9 MuiButton-outlinedPrimary"
        disabled={false}
        id="deleteTask"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          <span
            className="MuiButton-startIcon MuiButton-iconSizeMedium"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
              />
            </svg>
          </span>
          Delete task
        </span>
      </button>
    </div>
  </div>
</form>
`;
